File: core/src/main/java/heroes/journey/entities/actions/Action.java
package heroes.journey.entities.actions;

import static heroes.journey.registries.Registries.ActionManager;

import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Table;

import heroes.journey.entities.actions.inputs.ActionInput;
import heroes.journey.entities.actions.results.AIOnSelectNotFound;
import heroes.journey.entities.actions.results.ActionResult;
import heroes.journey.registries.Registrable;
import heroes.journey.ui.HUD;
import heroes.journey.ui.windows.InfoProvider;
import lombok.Getter;

public abstract class Action extends Registrable implements InfoProvider {

  @Getter protected final String description;
  @Getter protected final boolean returnsActionList;

  protected final Cost cost;

  public Action(String id, String name, String description, boolean returnsActionList, Cost cost) {
    super(id, name);
    this.description = description != null ? description : "";
    this.returnsActionList = returnsActionList;
    this.cost = cost != null ? cost : new Cost();
  }

  public Action(String id, String name) {
    this(id, name, "", false, null);
  }

  public ShowAction internalRequirementsMet(ActionInput input) {
    return ShowAction.YES;
  }

  public void internalOnHover(ActionInput input) {
  }

  public abstract ActionResult internalOnSelect(ActionInput input);

  public ActionResult internalOnSelectAI(ActionInput input) {
    return new AIOnSelectNotFound();
  }

  public ShowAction requirementsMet(ActionInput input) {
    return internalRequirementsMet(input).and(cost.requirementsMet(input));
  }

  public void onHover(ActionInput input) {
    HUD.get().getCursor().setMapPointerLoc(null);
    internalOnHover(input);
  }

  public ActionResult onSelect(ActionInput input, boolean ai) {
    if (input != null)
      cost.onUse(input);
    if (ai) {
      ActionResult aiResult = internalOnSelectAI(input);
      if (!(aiResult instanceof AIOnSelectNotFound)) {
        return aiResult;
      }
    }
    return internalOnSelect(input);
  }

  public ActionResult onSelect(ActionInput input) {
    return onSelect(input, false);
  }

  public Action register() {
    return ActionManager.register(this);
  }

  @Override
  public String getTitle() {
    return toString();
  }

  @Override
  public void fillCustomContent(Table table, Skin skin) {
    table.add(cost.getDisplay()).center().fill().expand();
  }
}

File: core/src/main/java/heroes/journey/entities/actions/Action.kt
package heroes.journey.entities.actions

import heroes.journey.GameState
import heroes.journey.modlib.actions.ActionResult
import heroes.journey.modlib.actions.IAction
import heroes.journey.modlib.actions.IActionContext
import heroes.journey.modlib.actions.ShowAction
import heroes.journey.modlib.attributes.Attributes
import heroes.journey.modlib.attributes.Stat
import heroes.journey.modlib.registries.InfoProvider
import heroes.journey.modlib.registries.Registrable
import heroes.journey.mods.Registries

open class Action(
    id: String,
    override val isReturnsActionList: Boolean = false,
    override val requirementsMetFn: (IActionContext) -> ShowAction = { ShowAction.YES },
    override val onHoverFn: (IActionContext) -> Unit = {},
    override val onSelectFn: (IActionContext) -> ActionResult,
    override val turnCooldown: Int = 0,
    override val factionCooldown: Boolean = false,
    override val cost: Attributes? = null,
    override val customInfoProviderFn: ((IActionContext) -> InfoProvider)? = null,
    override val requiredAllTags: List<Stat> = emptyList(),
    override val requiredAnyTags: List<Stat> = emptyList(),
    override val forbiddenTags: List<Stat> = emptyList()
) : Registrable(id), IAction {

    override fun getTitle(input: IActionContext): String {
        val customProvider = customInfoProviderFn?.invoke(input)
        if (customProvider != null) {
            return customProvider.getTitle(input)
        }
        return super.getTitle(input)
    }

    override fun getDescription(input: Map<String, String>): String {
        val actionContext = ActionContext(GameState.global(), null, false, input)
        val customProvider = customInfoProviderFn?.invoke(actionContext)
        if (customProvider != null) {
            return customProvider.getDescription(input)
        }
        return super.getDescription(input)
    }

    override fun register(): Action {
        return Registries.ActionManager.register(this) as Action
    }
}
